cpp/src/groupby/hash/groupby.cu

```diff
@@ -0,0 +1,481 @@
+ * upper bound of the output size).
+ *
+ * 2.) The final result is materialized by extracting the non-empty keys from
+ * the hash map and the non-empty rows from the sparse output aggregation table.
+ * Every non-empty key and value row is appended to the output key and value
+ * tables.
+ *
+ * @tparam keys_have_nulls Indicates keys have one or more null values
+ * @tparam values_have_nulls Indicates values have one or more null values
+ * @param keys Table whose rows are used as keys of the groupby
+ * @param values Table whose rows are aggregated in the groupby
+ * @param ops Set of aggregation operations to perform for each element in a row
+ * in the values table
+ * @param options Options to control behavior of the groupby operation
+ * @param stream CUDA stream on which all memory allocations and kernels will be
+ * executed
+ * @return A pair of the output keys table and output values table
+ *---------------------------------------------------------------------------**/
+template <bool keys_have_nulls, bool values_have_nulls>
+auto compute_hash_groupby(cudf::table const& keys, cudf::table const& values,
+                          std::vector<operators> const& ops, Options options,
+                          cudaStream_t stream) {
+  CUDF_EXPECTS(values.num_columns() == static_cast<gdf_size_type>(ops.size()),
+               "Size mismatch between number of value columns and number of "
+               "aggregations.");
+
+  std::vector<std::pair<gdf_column const*, operators>> requested_aggregations(
+      values.num_columns());
+
+  std::transform(values.begin(), values.end(), ops.begin(),
+                 requested_aggregations.begin(),
+                 [](gdf_column const* col, operators op) {
+                   return std::make_pair(col, op);
+                 });
+
+  cudf::table minimal_columns;
+  std::vector<operators> minimal_operators;
+  std::tie(minimal_columns, minimal_operators) =
+      compound_to_primitive(requested_aggregations);
+
+  auto const d_input_keys = device_table::create(keys);
+  auto const d_input_values = device_table::create(minimal_columns);
+
+  auto result = build_aggregation_map<keys_have_nulls, values_have_nulls>(
+      keys, values, *d_input_keys, *d_input_values, minimal_operators, options,
+      stream);
+
+  auto const map{std::move(result.first)};
+  cudf::table const sparse_output_values{result.second};
+
+  cudf::table output_keys;
+  cudf::table output_values;
+  //cudf::table minimal_output_values;
+  //std::tie(output_keys, minimal_output_values) =
+  std::tie(output_keys, output_values) =
+      extract_results<keys_have_nulls, values_have_nulls>(
+          keys, values, *d_input_keys, sparse_output_values, map.get(), stream);
+
+  //std::vector<std::pair<gdf_column*,operators>> computed_aggregations(minimal_output_values.num_columns());
+
+  //cudf::table output_values = minimal_to_final(minimal_output_values, );
```

Comment:
Unused code?

Reply:
Yeah, looks like you started reviewing before I'd pushed the final changes. You may want to re-look at the handling of compound/single aggregation requests.